Membership module refactoring

shamil-gadelshin opened the pull request:
### Changes
- incapsulate storage fields MemberIdsByRootAccountId and MemberIdsByControllerAccountId. Remove public access to those fields.
- rename Profile to Membership, and storage field MemberProfile to MembershipById. 
- rename MembersCreated => NextMemberId.
- rename Handles => MemberIdByHandle.
- drop UserInfo.
- CheckedUserInfo was renamed to the ValidatedUserInfo and kept for simplicity as a private internal structure.
- drop Option wrapper for around value types of maps
- remove roles from the membership module
- remove roles from the content working group module
- move ActorId from membership to the content working group
- add MemberId to the content working group Lead
- move Role from membership to roles module

shamil-gadelshin committed:
Rename MembersCreated to the NextMemberId

shamil-gadelshin committed:
Remove members mod from the membership

- move all the code from the members mod to the lib.rs
- update all dependencies

shamil-gadelshin committed:
Rename Handles to the MemberIdByHandle in the membership

shamil-gadelshin committed:
Rename Profile to Membership inte membership module

shamil-gadelshin committed:
Remove UserInfo - use its values as parameters in the memberhip module

shamil-gadelshin committed:
Make membership storage fields private

Make MemberIdsByRootAccountId and MemberIdsByControllerAccountId private in the membership module.

shamil-gadelshin committed:
Rename Memberships to the MembershipById

shamil-gadelshin committed:
Remove Option from MembershipById storage field

shamil-gadelshin committed:
Remove Option from the PaidMembershipTerms map

shamil-gadelshin committed:
Remove Option from ScreeningAuthority in the membership module

shamil-gadelshin committed:
Fix comments

shamil-gadelshin committed:
Remove roles from the membership module

- remove roles from the membership module
- remove roles from the content working group module
- move ActorId from membership to the content working group
- add MemberId to te content working group Lead
- update tests
- delete commented code

shamil-gadelshin committed:
Move Role from membership to roles module

bedeho commented on runtime-modules/membership/src/lib.rs:
```diff
@@ -1,9 +1,647 @@
+        /// Mapping of member's id to their membership profile
+        pub MembershipById get(membership) : map T::MemberId => MembershipOf<T>;
+
+        /// Mapping of a root account id to vector of member ids it controls.
```
These two fields were supposed to be blocked

https://github.com/Joystream/joystream/issues/93#issue-522940475

It's actually a substantial issue because there is non-trivial business logic here and in other modules that depend on the implied (controller & role) account uniqueness, hence removing them will help us track down and update all of those.

shamil-gadelshin replied:
I incapsulate them to be crate-only: the 'allowed check' is_member() requires them as main components. So, they are blocked. Also, I got rid of their getters, so no one can use this collections from external modules.

bedeho commented on runtime-modules/membership/src/lib.rs:
```diff
@@ -1,9 +1,647 @@
+    }
+
+    /// Returns true if account is either a member's root or controller account
+    pub fn is_member_account(who: &T::AccountId) -> bool {
```
This must go.

shamil-gadelshin replied:
This is the 'allowed check' from the https://github.com/Joystream/joystream/issues/93#issue-522940475 (task item 2).

bedeho replied:
Seeing that point now, sorry for the confusion this caused, this was probably meant to say something like `ensure_*(origin, MemberId)` is ok, but clearly its not. But we do not want this. There should be no constraint on the account space, as this requires, so `is_member_account` should go, and everything that relies on it in the runtime must be updated.

bedeho commented on runtime-modules/membership/src/lib.rs:
```diff
@@ -1,9 +1,647 @@
+    }
+
+    /// Ensure that given member has given account as the controller account
+    pub fn ensure_is_controller_account_for_member(
```
I think this is exclusively used for authenticating claimed members, both in this module, and by other caller module. We may as well make it `ensure_member_signed_as_controller(origin, member_id)`.

shamil-gadelshin replied:
Yes. And when I try to create it, I found a similar method with a slightly different name `ensure_member_controller_account_signed()` nearby.

bedeho replied:
Ok, so we remove this then?

shamil-gadelshin replied:
Removed that entry, but found another one more complicated use-case in the content working group in the apply_on_curator_opening(). So we keep it for now.

shamil-gadelshin commented:
Several issues were created in order to remove` is_member()`:
- https://github.com/Joystream/joystream/issues/519
- https://github.com/Joystream/joystream/issues/520
- https://github.com/Joystream/joystream/issues/521
- https://github.com/Joystream/joystream/issues/522

shamil-gadelshin committed:
Refactor content working group

Change call from ensure_is_controller_account_for_member() to the ensure_member_controller_account_signed().

shamil-gadelshin committed:
Apply rust fmt
